File: LeetCode Solutions/Study Plan/Programming Skills II/Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if(!head){
            return;
        }
        ListNode *temp = head, *half = head, *previous = NULL;
        while(temp->next && temp->next->next){
            temp = temp->next->next;
            half = half->next;
        }
        if(temp->next){
            half = half->next;
        }
        while(half){
            temp = half->next;
            half->next = previous;
            previous = half;
            half = temp;
        }
        half = previous;
        while(head && half){
            temp = head->next;
            previous = half->next;
            head->next = half;
            half->next = temp;
            head = temp;
            half = previous;
        }
        if(head && head->next){
            head->next->next = NULL;
        }
    }
};

File: LeetCode Solutions/Study Plan/Programming Skills/Programming Skills II/Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null){
            return;
        }
        ListNode temp = head, half = head, previous = null;
        while (temp.next != null && temp.next.next != null){
            temp = temp.next.next;
            half = half.next;
        }
        if (temp.next != null){
            half = half.next;
        }
        while(half != null){
            temp = half.next;
            half.next = previous;
            previous = half;
            half = temp;
        }
        half = previous;
        while (head != null && half != null){
            temp = head.next;
            previous = half.next;
            head.next = half;
            half.next = temp;
            head = temp;
            half = previous;
        }
        if (head != null && head.next != null){
            head.next.next = null;
        }
    }
}